Improve zfs receive performance with lightweight write

ahrens opened the pull request:
<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://openzfs.github.io/openzfs-docs/Developer%20Resources/Buildbot%20Options.html
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
The performance of `zfs receive` can be bottlenecked on the CPU consumed
by the `receive_writer` thread, especially when receiving streams with
small compressed block sizes.  Much of the CPU is spent creating and
destroying dbuf's and arc buf's, one for each `WRITE` record in the send
stream.

I described this work in my talk at the 2020 OpenZFS Developer Summit: [slides](https://docs.google.com/presentation/d/1HuKHawQbuetqpbwp4wmfm6Ozj-WYJpPa6QAxgDxLsgk/edit#slide=id.g9dcf4f23d1_0_2) (link should start you at slide 25) and [video](https://youtu.be/p9VHwuCRPes?t=1312) (link should start you at 21:52)

### Description
<!--- Describe your changes in detail -->
This commit introduces the concept of "lightweight writes", which allows
`zfs receive` to write to the DMU by providing an ABD, and instantiating
only a new type of `dbuf_dirty_record_t`.  The dbuf and arc buf for this
"dirty leaf block" are not instantiated.

Because there is no dbuf with the dirty data, this mechanism doesn't
support reading from "lightweight-dirty" blocks (they would see the
on-disk state rather than the dirty data).  Since the dedup-receive code
has been removed, `zfs receive` is write-only, so this works fine.

Because there are no arc bufs for the received data, the received data
is no longer cached in the ARC.

The code is also restructured in a few ways:

Added a `dr_dnode` field to the dbuf_dirty_record_t.  This simplifies
some existing code that no longer needs `DB_DNODE_ENTER()` and related
routines.  The new field is needed by the lightweight-type dirty record.

To ensure that the `dr_dnode` field remains valid until the dirty record
is freed, we have to ensure that the `dnode_move()` doesn't relocate the
dnode_t.  To do this we keep a hold on the dnode until it's zio's have
completed.  This is already done by the user-accounting code
(`userquota_updates_task()`), this commit extends that so that it always
keeps the dnode hold until zio completion (see `dnode_rele_task()`).

`dn_dirty_txg` was previously zeroed when the dnode was synced.  This
was not necessary, since its meaning can be "when was this dnode last
dirtied".  This change simplifies the new `dnode_rele_task()` code.

Removed some dead code related to `DRR_WRITE_BYREF` (dedup receive).

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->
<!--- Please think about using the draft PR feature if appropriate -->

Testing a receive of a stream with average compressed block size of 4KB,
this commit improves performance by 50%, while also reducing CPU usage
by 50% of a CPU.  On a per-block basis, CPU consumed by receive_writer()
and dbuf_evict() is now 1/7th (14%) of what it was.

Baseline: 450MB/s, CPU in receive_writer() 40% + dbuf_evict() 35%
New: 670MB/s, CPU in receive_writer() 17% + dbuf_evict() 0%

<img width="575" alt="image" src="https://user-images.githubusercontent.com/799124/97060483-4d154400-1548-11eb-89d3-9affd4cf9b8e.png">

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

adamdmoss commented on module/zfs/dmu.c:
```diff
@@ -1560,6 +1560,32 @@ dmu_return_arcbuf(arc_buf_t *buf)
+ * CPU cost of creating a dmu_buf_impl_t and arc_buf_[hdr_]_t.  However, the
+ * data can not be read or overwritten until the transaction's txg has been
+ * synced.  This makes it appropriate for workloads that are known to be
+ * (temporarily) write-only, like "zfs receive".
```
Can you think of workloads besides 'zfs receive' which are write-only, which this could apply to?  Resilvering perhaps...?

Although this change is pleasantly small, I get the impression that historically send+recv has been a hotspot for long-lurking bugs - so if the lightweight-write mechanism had other features that also used it I'd feel more confident that this was getting well-exercised.

Or a bunch more tests, I guess. :)

adamdmoss replied:
(Take the suggestion with a pinch of salt.  I'm just a dilletante. :) )

ahrens replied:
I'll keep an eye out for other write-only workloads.  Resilvering writes don't go through the DMU to begin with.

ahrens commented:
@behlendorf It looks like some of the ZTS runs are reporting memory leaks.  Do you have any tricks for debugging that?  E.g. can I get the stack traces where the leaked memory was allocated?

behlendorf commented:
Yes in fact you can.  The coverage builder's kmemleak checker will dump to the console log where the leaked memory was allocated.  I believe this is the stack you're looking for (extracted and cleanup from the following log).

http://build.zfsonlinux.org/builders/Ubuntu%2018.04%20x86_64%20Coverage%20%28TEST%29/builds/3522/steps/shell_9/logs/console

```
[15358.401767] RIP: 0010:strncpy+0x27/0x60
[15358.401767] RSP: 0018:ffffc900019fb5a0 EFLAGS: 00000297 ORIG_RAX: ffffffffffffff0c
[15358.401769] RAX: ffff8802c40a72ec RBX: 0000000000000040 RCX: 00000003ced7ceb7
[15358.401770] RDX: 00000003ced7cebb RSI: ffff8801ff7c615b RDI: ffff8802c40a72ec
[15358.401771] RBP: ffffc900019fb5a0 R08: 0000000000000000 R09: ffff87fef532a441
[15358.401772] R10: ffffffff82b96ab0 R11: ffff8803546dc000 R12: 0000000000000001
[15358.401772] R13: ffff8802c40a71d0 R14: ffff8803546dc000 R15: ffff8802c40a71d8
[15358.401773]  </IRQ>
[15358.401777]  create_object+0x103/0x380
[15358.401779]  kmemleak_alloc+0x7a/0x100
[15358.401780]  __kmalloc_node+0x25a/0x4e0
[15358.401803]  spl_kmem_alloc_impl+0xca/0x190 [spl]
[15358.402043]  spl_kmem_alloc+0x63/0x70 [spl]
[15358.402186]  zio_push_transform+0x41/0xa0 [zfs]
[15358.402326]  zio_read_bp_init+0x32c/0x410 [zfs]
[15358.402466]  zio_nowait+0x176/0x450 [zfs]
[15358.402584]  arc_read+0x1167/0x24a0 [zfs]
[15358.402820]  dbuf_read_impl.constprop.1+0x49b/0x760 [zfs]
[15358.402939]  dbuf_read+0x1ae/0x850 [zfs]
[15358.403181]  dmu_buf_hold+0x87/0xd0 [zfs]
[15358.403318]  zap_lockdir+0x42/0x140 [zfs]
[15358.403719]  zap_lookup_norm+0x4c/0xc0 [zfs]
[15358.403856]  zap_lookup+0x1e/0x30 [zfs]
[15358.403983]  dsl_dir_hold+0x17d/0x490 [zfs]
[15358.404108]  dsl_dataset_hold_flags+0x3e/0x410 [zfs]
[15358.404488]  dmu_objset_hold_flags+0x70/0x100 [zfs]
[15358.404870]  dmu_objset_hold+0x1e/0x30 [zfs]
[15358.405008]  zfs_destroy_unmount_origin+0x2c/0x120 [zfs]
[15358.406587]  dsl_destroy_head+0x22/0x200 [zfs]
[15358.406987]  zfs_ioc_create+0x268/0x580 [zfs]
[15358.407125]  zfsdev_ioctl_common+0x24d/0x8b0 [zfs]
[15358.407268]  zfsdev_ioctl+0x70/0x130 [zfs]
[15358.407272]  do_vfs_ioctl+0xa1/0x9c0
[15358.407276]  SyS_ioctl+0xad/0xe0
[15358.407278]  entry_SYSCALL_64_fastpath+0x1a/0xa5
```

ahrens commented:
I fixed the leak (which occurs when receiving an incremental large-block stream into a dataset that previously did a non-large-block receive), and addressed code review comments that I received out-of-band.

behlendorf commented on module/zfs/dbuf.c:
```diff
@@ -3984,6 +4044,153 @@ dbuf_sync_leaf_verify_bonus_dnode(dbuf_dirty_record_t *dr)
+		dmu_buf_impl_t *parent_db = dr->dr_parent->dr_dbuf;
+		int epbs = dn->dn_indblkshift - SPA_BLKPTRSHIFT;
+		ASSERT3U(parent_db->db_level, ==, 1);
+		ASSERT3U(parent_db->db_dnode_handle->dnh_dnode, ==, dn);
```
```suggestion
		ASSERT3P(parent_db->db_dnode_handle->dnh_dnode, ==, dn);
```

```
../../module/zfs/dbuf.c: In function 'dbuf_lightweight_bp':
../../lib/libspl/include/assert.h:83:26: error: cast from pointer to integer of different size [-Werror=pointer-to-int-cast]
  const uint64_t __left = (uint64_t)(LEFT);   \
```

ahrens replied:
thanks, fixed!

behlendorf commented:
@grwilson @pcd1193182 if possible would you mind wrapping up your reviews of this PR. Thanks.

ahrens commented:
Rebased.  @grwilson @pcd1193182 could you take a look at this?

grwilson commented on module/zfs/dbuf.c:
```diff
@@ -3984,6 +4044,153 @@ dbuf_sync_leaf_verify_bonus_dnode(dbuf_dirty_record_t *dr)
+	dnode_t *dn = dr->dr_dnode;
+
+	if (dn->dn_phys->dn_nlevels == 1) {
+		ASSERT3U(dr->dt.dll.dr_blkid, <, dn->dn_phys->dn_nblkptr);
+		return (&dn->dn_phys->dn_blkptr[dr->dt.dll.dr_blkid]);
+	} else {
+		dmu_buf_impl_t *parent_db = dr->dr_parent->dr_dbuf;
+		int epbs = dn->dn_indblkshift - SPA_BLKPTRSHIFT;
+		ASSERT3U(parent_db->db_level, ==, 1);
+		ASSERT3P(parent_db->db_dnode_handle->dnh_dnode, ==, dn);
+		ASSERT3U(dr->dt.dll.dr_blkid >> epbs, ==, parent_db->db_blkid);
```
Does it make sense to convert some of these to VERIFYs?

ahrens replied:
Sure, I'll change all of these.

pcd1193182 commented on module/zfs/dbuf.c:
```diff
@@ -1973,6 +1973,75 @@ dbuf_redirty(dbuf_dirty_record_t *dr)
+dbuf_dirty_lightweight(dnode_t *dn, uint64_t blkid, dmu_tx_t *tx)
+{
+	rw_enter(&dn->dn_struct_rwlock, RW_READER);
+	ASSERT(!dn->dn_objset->os_raw_receive ||
```
[nit] personally I feel like IMPLY(os_raw_receive, maxblkid >= blkid) is slightly clearer than this?

ahrens replied:
I agree, changed.
